src/Common/ShellCommand.cpp

```diff
@@ -160,64 +165,93 @@ std::unique_ptr<ShellCommand> ShellCommand::executeImpl(
     for (size_t i = 0; i < config.write_fds.size(); ++i)
         write_pipe_fds.emplace_back(std::make_unique<PipeFDs>());
 
-    pid_t pid = reinterpret_cast<pid_t(*)()>(real_vfork)();
+    pid_t child_pid = reinterpret_cast<pid_t(*)()>(real_vfork)();
 
-    if (pid == -1)
+    if (child_pid == -1)
         throw ErrnoException(ErrorCodes::CANNOT_FORK, "Cannot vfork");
 
-    if (0 == pid)
+    /// Here we are using double vfork technique to prevent zombie process.
+    if (0 == child_pid)
     {
-        /// We are in the freshly created process.
-
-        /// Why `_exit` and not `exit`? Because `exit` calls `atexit` and destructors of thread local storage.
-        /// And there is a lot of garbage (including, for example, mutex is blocked). And this can not be done after `vfork` - deadlock happens.
+        pid_t pid = reinterpret_cast<pid_t(*)()>(real_vfork)();
```

Comment:
It may not be completely safe.([See `Standard description` section](https://man7.org/linux/man-pages/man2/vfork.2.html)).

But we're already doing memory modification, so suppose we're not doing anything worse.